# 4.3 Classes in Python 

## Objects

Using functions is the most important way to write scientific
code. The basic approach is to have blocks of code that take in data
and return results; this is called *procedural programming*. But there
is also another way in which data and functions are combined into
something called an *object*, which leads to
[object oriented programming](https://python.swaroopch.com/oop.html)
(OOP). An object contains data (held in variables that are called
*attributes*) and it also contains functions (called *methods*) that
know how to operate on the data in the object.

Python is an *object oriented* (OO) language and objects are
everywhere --- in fact *everything* is an object in Python.

### Some Python objects
Even if you don't use object-oriented programming, you still need to
know how to work with Python objects. We look at a few examples.

Each [built-in type](https://docs.python.org/3/library/stdtypes.html)
(`int`, `float`, `str`, ...) is an object with associated *methods*
and *attributes*.

#### Strings
The text sequence --- or "string" --- type
[`str`](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str)
has [lots of string
methods](https://docs.python.org/3/library/stdtypes.html#string-methods):

In [1]:
sentence = "may the force be with you!"

In [2]:
sentence.capitalize()

'May the force be with you!'

In [3]:
sentence.upper()

'MAY THE FORCE BE WITH YOU!'

In [4]:
sentence.count("o")

2

In [5]:
sentence.isdigit()

False

In [6]:
sentence.replace("you", "us")

'may the force be with us!'

In [7]:
sentence.split()

['may', 'the', 'force', 'be', 'with', 'you!']

Note that you don't have to assign a string to a variable: the _string object itself contains the methods_:

In [8]:
"may the force be with you!".upper()

'MAY THE FORCE BE WITH YOU!'

The output of many of these methods is again a string so one can
easily concatenate or "chain" methods:

In [9]:
sentence.replace("you", "us").title()

'May The Force Be With Us!'

#### Lists
The [`list`](https://docs.python.org/3/library/stdtypes.html#lists)
type contains a [large number of useful
methods](https://docs.python.org/3/library/stdtypes.html#mutable-sequence-types)
that allow one to manipulate the list. Typically, all operations are
done "in place", i.e., they change the list itself.


In [10]:
rebels = ["Luke", "Leia", "Han", "Chewie"]
print(rebels)

['Luke', 'Leia', 'Han', 'Chewie']


In [11]:
rebels.append("Lando")
print(rebels)

['Luke', 'Leia', 'Han', 'Chewie', 'Lando']


In [12]:
rebels.pop()

'Lando'

In [13]:
print(rebels)

['Luke', 'Leia', 'Han', 'Chewie']


In [14]:
rebels.remove("Han")
print(rebels)

['Luke', 'Leia', 'Chewie']


In [15]:
rebels.extend(["R2D2", "C3PO"])

In [16]:
print(rebels)

['Luke', 'Leia', 'Chewie', 'R2D2', 'C3PO']


In [17]:
rebels.reverse()
print(rebels)

['C3PO', 'R2D2', 'Chewie', 'Leia', 'Luke']


In [18]:
rebels.sort()
print(rebels)

['C3PO', 'Chewie', 'Leia', 'Luke', 'R2D2']


In [19]:
rebels.insert(2, "Han")
print(rebels)

['C3PO', 'Chewie', 'Han', 'Leia', 'Luke', 'R2D2']


In [20]:
rebels.clear()
print(rebels)

[]


## Creating objects: classes

In Python one creates an object by first defining a
[class](https://docs.python.org/3/tutorial/classes.html#a-first-look-at-classes):


In [21]:
import math

class Sphere:
    """A simple sphere."""
 
    def __init__(self, pos, radius=1):
        self.pos = tuple(pos)
        self.radius = float(radius)

    def volume(self):
        return 4/3 * math.pi * self.radius**3

    def translate(self, t):
        self.pos = tuple(xi + ti for xi, ti in zip(self.pos, t))


and then *instantiating* the object (creating an instance of the class)

In [22]:
ball = Sphere((0, 0, 10), radius=2)

Notes on the class definition above:

* `__init__()` is a special [method](#attributes-and-methods) that is
  called when the class is instantiated: it is used to populate the
  object with user-defined values.
* The first argument of each [method](#attributes-and-methods)
  (including `__init__()`) is always
  called `self` and refers to the class itself.
* So-called instance [attributes](#attributes-and-methods) are created with
  `self.ATTRIBUTE_NAME`, e.g., `self.pos`.
* [Methods](#attributes-and-methods) are defined just like ordinary
  Python [functions](#functions) except that the first argument is `self`.


In this example we created an object named `ball`, which is of type
`Sphere`:

In [23]:
 type(ball)

__main__.Sphere

Note: It is convenient to put the class definition into a separate module, let's say `bodies.py`, and then you can import the class definitions as
```python
from bodies import Sphere
```
This tends to be more manageable than working interactively and it is an excellent way to modularize code.

## Attributes and Methods

Objects contain **attributes** (variables that are associated with the
object) and **methods** (functions that are associated with the
object). Attributes and methods are accessed with the "dot"
operator. (Within the *class definition*, attributes and methods are
also accessed with the dot-operator but the class itself is referred
to as `self` --- this is just the first argument in each method and
*you should always, always name it "self"*.)

In the example, `pos` and `radius` are
attributes, and can be accessed as `ball.pos` and `ball.radius`. For
instance, the `Sphere` object named `ball` has position


In [24]:
ball.pos

(0, 0, 10)

In [25]:
ball.radius

2.0

because we provided the `pos` argument `(0, 0, 10)` on
instantiation. Similarly, we created a ball with radius 2.


One can assign to these attributes as usual, e.g., directly change the position 

In [26]:
ball.pos = (-5, 0, 0)
ball.pos

(-5, 0, 0)

The `Sphere.volume()` method computes the volume of the sphere:

In [27]:
ball.volume()

33.510321638291124

Note that even though the method has the first argument `self` in its definition, we do not provide it when the method is called! This is special behavior for methods.

The `Sphere.translate()` method changes the position of the object by
adding a translation vector `t` to `Sphere.pos`:


In [28]:
ball.translate((5, 0, 0))
ball.pos

(0, 0, 0)

Note that this method did not return any values but it changed the
data in `Sphere.pos`.

## Independence of instances

Each instance of a class is independent from the other instances. For
example, `ball` and a new `balloon` can be moved independently
even though we start them at the same position:

In [29]:
ball = Sphere((0, 0, 10), radius=2)
balloon = Sphere((0, 0, 10), radius=6)

ball.pos = (-1, -1, 0)

In [30]:
print(ball.pos, balloon.pos)

(-1, -1, 0) (0, 0, 10)


## Inheritance

New classes can be built on existing classes in such a way that the
new class contains the functionality of the existing class. This is
called *inheritance* and is a very powerful way to organize large code
bases.

Only a small example is given to illustrate the basic idea: We use our
`Sphere` class to create planets. A planet is (almost) a sphere but it
also has a name and a mass: The new `Planet` class inherits from
`Sphere` by supplying `Sphere` as an argument to `Planet`:


In [31]:
class Planet(Sphere):
    def __init__(self, name, pos, mass, radius):
        self.name = str(name)
        self.pos = tuple(pos)
        self.mass = float(mass)
        self.radius = float(radius)

    def density(self):
        """Compute density of the planet"""
        return self.mass / self.volume()

In [32]:
# quantities from http://www.wolframalpha.com
# lengths in m and mass in kg
earth = Planet("Earth", (1.4959802296e11 , 0, 0), 5.9721986e24, 6371e3)
print(earth.density())

5513.442083060875


gives 5513 kg/m<sup>3</sup> because the `Planet` class inherited the
`volume()` method from `Sphere`.